### DISCUSSION TITLE
Remove "y" and "z" from secret generation

### DISCUSSION THREAD

**Discussion by gerhard1712** [pretix/pretix#3791] [Ideas]
When use web-checkin with a qr code scanner it paste the scanned qr code into the input field. But when use german keyboard layout, it is a qwertz-layout instead of a qwerty-layout. This results to the problem, that the scanned code is not valid when it includes a 'y' or 'z'. The workaround here is to change the keyboard layout in the operating system.

`class RandomTicketSecretGenerator(BaseTicketSecretGenerator):
    verbose_name = _('Random (default, works with all pretix apps)')
    identifier = 'random'
    use_revocation_list = False

    def generate_secret(self, item: Item, variation: ItemVariation = None, subevent: SubEvent = None,
                        attendee_name: str = None, valid_from: datetime = None, valid_until: datetime = None,
                        current_secret: str = None, force_invalidate=False) -> str:
        if current_secret and not force_invalidate:
            return current_secret
        return get_random_string(
            length=self.event.settings.ticket_secret_length,
            # Exclude o,0,1,i,l to avoid confusion with bad fonts/printers
            allowed_chars='abcdefghjkmnpqrstuvwxyz23456789'
        )`

The allowed chars are set here. 
When remove also 'y' and 'z' as allowed chars this behavior wouldn't occur.

**Comment by raphaelm** [comment]
Hmm, yes, good point, this is one of the most common problems with pretixscan desktop, although then we have solved it for Germany, but not for France or other countries and before we know it we barely have characters left at all ;)

> The workaround here is to change the keyboard layout in the operating system.

That's the workaround, yes, whereas the recommended solution is to just reconfigure your barcode scanner to qwertz. I have not yet seen a barcode scanner that can't be reconfigured to a German layout.

**Reply by gerhard1712** [reply]
Thanks for the fast answer.
Is it possible to make the allowed_chars configurable via settings?
It is already used for secret length.
So the "o,0,1,i,l" can added when use qr codes, because in this usage no typing is needed. And the problem with qwertz and qwerty can also fixed by each organizator.

**Reply by raphaelm** [reply]
I don't think we should add a setting for this, as it may be confusing to users and does not really make the situation easier for unexperienced users:

- An experienced organizer will have barcode scanners with correct configuration.
- An unexperienced organizer will not see this issue coming until it is already to late – there is no point in changing such configuration during your event when the QR codes have already been issued.

If you *really* want to, you could do this in a plugin, but I don't see why you wouldn't just fix the barcode scanner to emit correct data :)

> So the "o,0,1,i,l" can added when use qr codes, because in this usage no typing is needed. 

The "easy to read and type" support is intended as a backup-plan if QR codes are unreadable, e.g. because someone printed them on a really bad printer. I don't think this is a situation that can safely be assumed to never happen by an organizer.
